try, throw and catch

exception with arguments
```cpp
void divide (int a, int b) {
	if (b <= 0) throw b;
	else 
		return a/b;
}

int main() {
	try {
		divide(10, 0)
	}
	catch(int b) {
		std::cerr << 
	}
}
```

catch any exception
```cpp
try {
	// somethign
	throw // something
}
catch(...) {//everthing
	std::cerr << "Error!!";
}
```

making our own exception
```cpp
class exception {
	private:
		std::string m_error;
	public:
		exception(std::string error) : m_error{error} {}

		const std::string& getError() const {return m_error;}
}

class IntArray {
	private:
		int m_data[3] {};
	public:
		IntArray() {}
		int& operator[](const int index) {
			if (index < 0) 
				throw exception{'Out of bound'};
		}
}

int main() {
	IntArray array;
	try {
		int value {array[5]};
	}
	catch(const exception& exception) {
		std::cerr << exception.getError();
	}
		
}

THE OUTPUT : Out of bound
```

Function Templates:
```cpp
template<typename T, typename U>
T function_name(T a, U b) {
	if (a < b) 
		return a
	else 
		return b
}

int main() {
	int result = function_name<int, char>(10, 's');
}
```

Class Template:
```cpp
template<typename T, typename U>
class Something {
	private:
		T value {};
		U name {};
	public:
		void print() {
			std::cout << value << name << '\n';
		}
}

int main() {
	Something<int, std::string> a{10, 'something'};
	a.print();
}
```

Inheritance using class template:
```cpp
template<typename T>
class Base {}:

class Derived : public Base<int>{}; // initalize base as int
```

```cpp
template<typename T>
class Base{};

template<typename T> // this is used by the derived 
class Dervied : public Base<int> {}; // int is used by base
```

```cpp
template<typename T>
class Base{};

template<typename T> // T is used by both derived and base
class Dervied : public Base<T> {}; // 
```

```cpp
template<typename T>
class Base{};

template<typename T, typename U> // T is used by dervied and U is used by base
class Dervied : public Base<U> {};
```

```cpp
class Base{};

template <typename T>
class Dervied : public Base {}; // T not used by base and only used by derived
```